client/notices.go

```diff
@@ -0,0 +1,200 @@
+	"time"
+)
+
+type NotifyOptions struct {
+	// Key is the custom notice's key. Must be in "domain.com/key" format.
+	Key string
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+
+	// RepeatAfter allows the notice to repeat after this duration. If this is
+	// zero (the default), the notice will always repeat.
+	RepeatAfter time.Duration
+}
+
+// Notify records an occurrence of a "custom" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "custom",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Types, if not empty, includes only notices whose type is one of these.
+	Types []NoticeType
+
+	// Keys, if not empty, includes only notices whose key is one of these.
+	Keys []string
+
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
+	ID            string            `json:"id"`
+	Type          NoticeType        `json:"type"`
+	Key           string            `json:"key"`
+	FirstOccurred time.Time         `json:"first-occurred"`
+	LastOccurred  time.Time         `json:"last-occurred"`
+	LastRepeated  time.Time         `json:"last-repeated"`
+	Occurrences   int               `json:"occurrences"`
+	LastData      map[string]string `json:"last-data,omitempty"`
+	RepeatAfter   time.Duration     `json:"repeat-after,omitempty"`
+	ExpireAfter   time.Duration     `json:"expire-after,omitempty"`
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when the change is first spawned
+	// or its status is updated.
+	ChangeUpdateNotice NoticeType = "change-update"
+
+	// A custom notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	CustomNotice NoticeType = "custom"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	WarningNotice NoticeType = "warning"
```

Comment:
Seems a bit premature considering we have no supporting code for this, and have no timeframing for supporting it either. Same for ChangeUpdateNotice above: I think we should add the types together with the supporting code.

Reply:
Yeah, fair enough -- removed.